lib/handlers/events/delivery/goals/k8s/KubernetesGoalLauncher.ts

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    automationClientInstance,
+    AutomationContextAware,
+    configurationValue,
+    doWithRetry,
+    logger,
+} from "@atomist/automation-client";
+import {
+    ExecuteGoalResult,
+    GoalInvocation,
+    GoalLauncher,
+} from "@atomist/sdm";
+import * as k8s from "@kubernetes/client-node";
+import * as cluster from "cluster";
+import * as _ from "lodash";
+import * as os from "os";
+import { loadKubeConfig } from "./k8config";
+
+/**
+ * GoalLauncher implementation that runs SDM goals inside k8s jobs.
+ *
+ * It reuses the podSpec of the deployed SDM to create a new jobSpec from.
+ * Subclasses may change the spec and job creation behavior by overwriting beforeCreation
+ * and/or afterCreation methods.
+ */
+export class KubernetesGoalLauncher implements GoalLauncher {
+
+    constructor(private readonly options: { isolateAll?: boolean } = { isolateAll: false }) {
+        this.init();
+    }
+
+    public async supports(gi: GoalInvocation): Promise<boolean> {
+        return !process.env.ATOMIST_ISOLATED_GOAL &&
+            (
+                // Goal is marked as isolated and SDM is configured to use k8s jobs
+                (gi.goal.definition.isolated && process.env.ATOMIST_GOAL_LAUNCHER === "kubernetes") ||
+                // Force all goals to run isolated via env var
+                process.env.ATOMIST_GOAL_LAUNCHER === "kubernetes-all" ||
+                // Force all goals to run isolated via explicit option
+                (this.options.isolateAll && process.env.ATOMIST_GOAL_LAUNCHER === "kubernetes") ||
+                // Force all goals to run isolated via explicit configuration
+                _.get(gi.configuration, "sdm.kubernetes.isolateAll", false) === true
+            );
+    }
+
+    public async launch(gi: GoalInvocation): Promise<ExecuteGoalResult> {
+        const { goalEvent, context } = gi;
+
+        // Using new ATOMIST_POD_NAME to overwrite the host name default
+        // This is to prevent breakage when users still have old ATOMIST_DEPLOYMENT_NAME env var defined
+        const podName = process.env.ATOMIST_POD_NAME || os.hostname();
+        const podNs = process.env.ATOMIST_DEPLOYMENT_NAMESPACE || "default";
```

Comment:
Seems if we are using `ATOMIST_POD_NAME` we should also accept `ATOMIST_POD_NAMESPACE`.

Reply:
👍